src/ledger/tests.zig

```diff
@@ -425,3 +487,130 @@ pub fn TestDB(scope: []const u8) type {
+        _ = try inserter.insertShreds(shreds, is_repairs, null, false, null);
+        return timer.read();
+    }
+
+    // Analogous to [bench_read_sequential]https://github.com/anza-xyz/agave/blob/cfd393654f84c36a3c49f15dbe25e16a0269008d/ledger/benches/blockstore.rs#L78
+    pub fn benchReadSequential() !u64 {
+        const allocator = std.heap.c_allocator;
+        var state = try State.initBench("bentch read sequential");
+        defer state.deinitBench();
+        var inserter = try state.shredInserter();
+        var reader = try state.reader();
+
+        const prefix = "agave.blockstore.bench_read.";
+        const shreds = try benchShreds(prefix ++ "shreds.bin");
+        defer inline for (.{shreds}) |slice| {
+            deinitShreds(allocator, slice);
+        };
+
+        const total_shreds = shreds.len;
+
+        _ = try ledger.insert_shred.insertShredsForTest(&inserter, shreds);
+
+        const slot: u32 = 0;
+        const num_reads = total_shreds / 15;
+
+        var rng = std.Random.DefaultPrng.init(100);
+
+        var timer = try std.time.Timer.start();
+        const start_index = rng.random().intRangeAtMost(u32, 0, @intCast(total_shreds));
+        for (start_index..start_index + num_reads) |i| {
+            const shred_index = i % total_shreds;
+            _ = try reader.getDataShred(slot, shred_index);
+        }
+        return timer.read();
+    }
+
+    // Analogous to [bench_read_random]https://github.com/anza-xyz/agave/blob/92eca1192b055d896558a78759d4e79ab4721ff1/ledger/benches/blockstore.rs#L103
+    pub fn benchReadRandom() !u64 {
+        const allocator = std.heap.c_allocator;
+        var state = try State.initBench("bench read randmom");
+        defer state.deinitBench();
+        var inserter = try state.shredInserter();
+        var reader = try state.reader();
+
+        const prefix = "agave.blockstore.bench_read.";
+        const shreds = try benchShreds(prefix ++ "shreds.bin");
+        defer inline for (.{shreds}) |slice| {
+            deinitShreds(allocator, slice);
+        };
+
+        const total_shreds = shreds.len;
+        _ = try ledger.insert_shred.insertShredsForTest(&inserter, shreds);
+
+        const slot: u32 = 0;
+
+        var rng = std.Random.DefaultPrng.init(100);
+
+        var indices = try std.ArrayList(u32).initCapacity(inserter.allocator, total_shreds);
+        defer indices.deinit();
+        for (total_shreds) |_| {
+            indices.appendAssumeCapacity(rng.random().uintAtMost(u32, @intCast(total_shreds)));
+        }
+
+        var timer = try std.time.Timer.start();
+        for (indices.items) |shred_index| {
+            _ = try reader.getDataShred(slot, shred_index);
+        }
+        return timer.read();
+    }
+
+    // Analogous to [bench_serialize_write_bincode](https://github.com/anza-xyz/agave/blob/9c2098450ca7e5271e3690277992fbc910be27d0/ledger/benches/protobuf.rs#L88)
+    pub fn benchSerializeWriteBincode() !u64 {
```

Comment:
I am not sure if this is 100% identical to`bench_serialize_write_bincode` in agave has Agave has `put_bytes` that first serialized the values before insertion. `Sig` currently does not have similar method.

Reply:
In this case, the agave benchmark includes both serialization and insertion. So I think it is actually the same

Side note, regarding put_bytes not existing in sig. I tried to abstract away the bytes. Converting between types and bytes is the whole purpose of the Database layer. Ideally the ledger code can just use a Database like it's a hashmap, without worrying about serialization. It feels like a leaky abstraction if the Database layer can be bypassed, exposing its internal data to the business logic of the ledger.

That said, actually we do have a getBytes method, so I didn't do a great job at my goal haha. This is because shreds themselves are already bytes. The only reason getBytes exists is to provide a mechanism for managing the lifetime of the shred bytes without needing to copy them into a new allocation. I think this could actually be handled better using the type system, and we could probably eliminate getBytes completely. I decided was willing to compromise on the abstraction a bit, to get the job done quickly.

Reply:
>  think this could actually be handled better using the type system, and we could probably eliminate getBytes completely. I decided was willing to compromise on the abstraction a bit, to get the job done quickly.

Having a TODO could help with the rationale and also to have the intention to improve close to mind

Reply:
https://github.com/orgs/Syndica/projects/2/views/10?pane=issue&itemId=83834832